Add new cosa cloud-prune command for garbage collection

gursewak1997 opened the pull request:
Add the ability to run garbage collection on resources using
cosa cloud-prune. This script would take policy.yaml and run
the garbage collection accordingly for the stream specified.

---

Test it with:
`cosa cloud-prune --stream next-devel --gcp-json-key gcp-account-demo.json --gcp-project fedora-coreos-*** s3 --aws-config-file ./creds fcos-developer-pipelines/gursewak/next-devel
`

dustymabe commented on src/remote_prune:
```diff
@@ -0,0 +1,236 @@
+    if "cloud-uploads" in actions.keys():
+        cloud_uploads_duration = actions["cloud-uploads"]
+        build_duration = actions["build"]
+        # assumption we are keeping the duration in years
```
probably easier if we don't make assumptions.. i.e. maybe we just use a unit of days and even though it's a little harder to work with from a human perspective, it's easier not to make a mistake on the code side.

gursewak1997 replied:
Do we want to keep one unit i.e. could be just days or months or years? OR do we want to handle every possible inputs such as 90d, 4m and 2y?

dustymabe commented on src/remote_prune:
```diff
@@ -0,0 +1,236 @@
+#!/usr/bin/python3 -u
+
```
Let's define the format of the policy.yaml somewhere in this file? i.e. a nice description at the top of this file might be a good start

dustymabe commented on src/remote_prune:
```diff
@@ -0,0 +1,236 @@
+
+                            images = {
+                                "amis": metaData.get("amis") or [],
+                                "azure": metaData.get("azure") or [],
```
for FCOS at least we don't have any cloud resources for Azure that we need to cleanup (yet).

For RHCOS I think we do have some object storage uploads we do.

dustymabe commented on src/remote_prune:
```diff
@@ -0,0 +1,236 @@
+                except Exception as e:
+                    errors.append(e)
+
+    if len(errors) != 0:
+        print(f"Found errors when removing build {build.id}:")
+        for e in errors:
+            print(e)
```
should we actually raise an exception here instead of just being informational?

dustymabe commented on src/remote_prune:
```diff
@@ -0,0 +1,236 @@
+                    if build_id == "35.20210508.91.0":
+                        for arch in build["arches"]:
+                            print(f"Pruning {years} old {action} for {build_id} for {arch} in {stream}")
+                            # buildFetch(args.stream, build_id, arch)
```
needed?

gursewak1997 replied:
It is needed to get the updated meta.json. Skipping it while I run the testing on single build.

dustymabe commented on src/remote_prune:
```diff
@@ -0,0 +1,236 @@
+
+from cosalib.aws import deregister_ami, delete_snapshot
+
+Build = collections.namedtuple("Build", ["id", "timestamp", "images", "arches"])
```
note that we do have a Build() class in cosalib as well. Not sure if that would be useful here to use that instead or not.

gursewak1997 replied:
We do indeed have that but they are changed according to those files. It seemed better to create a new variable since we are using `cloud_images` here.

dustymabe commented on src/remote_prune:
```diff
@@ -0,0 +1,236 @@
+                                case "build":
+                                    print(f"Will delete everything for {build_id}")
+        # pprint.pprint(buildJsonData['builds'], compact=True)
+        with open(f"builds/builds.json", "w") as json_file:
+            json_file.write(json.dumps(buildJsonData))
```
just pointing out we will need some special handling of builds.json when we upload it back to s3

jlebon commented on src/remote_prune:
```diff
@@ -0,0 +1,224 @@
+        if dry_run:
+            print(f"Will delete {ami_id} and {snapshot_id}")
+        else:
+            if ami_id and region_name:
+                try:
+                    deregister_ami(ami_id, region=region_name)
+                except Exception as e:
+                    errors.append(e)
+            if snapshot_id and region_name:
+                try:
+                    delete_snapshot(snapshot_id, region=region_name)
+                except Exception as e:
+                    errors.append(e)
+
+    aliyun = build.images.get("aliyun")
+    if aliyun:
+        region_name = aliyun.get("name")
+        aliyun_id = aliyun.get("hvm")
+        if dry_run:
+            print(f"Will delete {aliyun_id} in {region_name}")
+        else:
+            if region_name and aliyun_id:
+                try:
+                    remove_aliyun_image(aliyun_id, region=region_name)
+                except Exception as e:
+                    errors.append(e)
+
+    gcp = build.images.get("gcp")
+    if gcp:
+        gcp_image = gcp.get("image")
+        json_key = cloud_config.get("gcp", {}).get("json-key")
+        project = cloud_config.get("gcp", {}).get("project")
+        if dry_run:
+            print(f"Will delete {gcp_image}")
+        else:
+            if gcp_image and json_key and project:
+                try:
+                    remove_gcp_image(gcp_image, json_key, project)
+                except Exception as e:
+                    errors.append(e)
```
Hmm, one concern here is designing this in a [fail-safe](https://en.wikipedia.org/wiki/Fail-safe) way. So if a new upload type is added and we forget to add support for it in this script, after long enough it could happen that we start deleting those builds and leak those resources. Ideally, this command should fail instead but that would require it to know about clouds it doesn't know about. :)

One straightforward but awkward way to do this is to have an allowlist of all toplevel keys allowed in the `meta.json`. If a toplevel key is in `meta.json` that we don't know about, we fail. Not sure if there's a less awkward way to do this. Ideally, there'd be a way to leverage the schema to help with this.

For example, right now this doesn't handle `azure`, `kubevirt`, `extensions-container`, and `base-oscontainer` (the latter three could have versioned tags that should ideally be pruned). And that'd be OK for really old builds that don't even have those. But without something to cross-check against, there's nothing that stops this code from also deleting builds that do have them.

jlebon replied:
OK, so here's my idea for this: maintain a list of supported (e.g. `amis`) and unsupported keys (e.g. `kubevirt`). The union of these two lists should represent the full set of possible remote resources that a build could have as per the schema.

If a build is up for cloud pruning, we check that none of the unsupported keys are present. If any are, error out.

To ensure that the list remains in sync with the schema, we keep a hash of the schema JSON file as a comment in this file, and add a CI check in this repo that greps for the hash of the schema. This means that any schema change requires changing the comment too. The comment could say something like:

```python
# These lists are up to date as of schema hash
# 4c19aed3b3d84af278780bff63728510bb3e70613e4c4eef8cabd7939eb31bd8. If changing
# this hash, ensure that the list of supported and unsupported artifacts below
# is up to date.
def supported = [...]
def unsupported = [...]
```

jlebon commented on src/remote_prune:
```diff
@@ -0,0 +1,198 @@
+    supported = ["amis"]
+    unsupported = ["base-oscontainer", "aliyun", "azure", "azurestack", "digitalocean", "exoscale", "gcp", "kubevirt", "hyperv", "ibmcloud"]
+
+    if stream in policy:
```
```suggestion
    if stream not in policy:
        return
    ... # dedent
```

jlebon commented on src/remote_prune:
```diff
@@ -0,0 +1,198 @@
+        # If the build key is set in the policy file, then the cloud-uploads key must
+        # also be present, and the duration of cloud-uploads must be equal or shorter
+        if "build" in policy[stream].keys():
+            cloudUploadsCheck(policy[stream])
```
This is minor but I think the prevailing naming convention in Python for functions/variable names is snake_case.

jlebon commented on src/remote_prune:
```diff
@@ -0,0 +1,198 @@
+        # also be present, and the duration of cloud-uploads must be equal or shorter
+        if "build" in policy[stream].keys():
+            cloudUploadsCheck(policy[stream])
+        BASE_URL = "https://builds.coreos.fedoraproject.org/prod/streams/" + stream
```
We should be fetching via S3 instead. This will make it compatible with RHCOS too and if the bucket isn't public.

I.e. the command should take as a parameter the `s3://` builds directory similarly to `cosa buildfetch`.

jlebon commented on src/remote_prune:
```diff
@@ -0,0 +1,198 @@
+        if "build" in policy[stream].keys():
+            cloudUploadsCheck(policy[stream])
+        BASE_URL = "https://builds.coreos.fedoraproject.org/prod/streams/" + stream
+        buildJsonData = getBuildsForStream(stream)
```
Let's not use `cosa buildfetch` for this. We don't need most of the functionality of `buildfetch` because we're operating across a wide array of builds, not just e.g. the latest. I think fetching `builds.json` directly ourselves will be cleaner.

jlebon commented on src/remote_prune:
```diff
@@ -0,0 +1,198 @@
+        # action is basically whatever is needed to be pruned for the respective stream
+        for action in policy[stream]:
+            duration = policy[stream][action]
+            if duration.split(" ")[1] == ("years" or "year"):
+                refDate = datetime.datetime.now() - relativedelta(years=int(duration.split(" ")[0]))
+            elif duration.split(" ")[1] == ("months" or "month"):
```
This syntax doesn't seem valid. I think you want e.g. `if unit in ['year', 'years']:`

jlebon commented on src/remote_prune:
```diff
@@ -0,0 +1,198 @@
+        # action is basically whatever is needed to be pruned for the respective stream
+        for action in policy[stream]:
+            duration = policy[stream][action]
+            if duration.split(" ")[1] == ("years" or "year"):
+                refDate = datetime.datetime.now() - relativedelta(years=int(duration.split(" ")[0]))
+            elif duration.split(" ")[1] == ("months" or "month"):
+                refDate = datetime.datetime.now() - relativedelta(months=int(duration.split(" ")[0]))
```
This is splitting the same string multiple times. You can do e.g.

```
val, unit = duration.split(" ")
```

upfront

jlebon commented on src/remote_prune:
```diff
@@ -0,0 +1,198 @@
+
+            print(f"Pruning {duration} old {action} for {stream} builds")
+            # Enumerating in reverse to go from the oldest build to the newest one
+            for index, build in reversed(list(enumerate(buildJsonData["builds"]))):
```
You can `enumerate` last so that we avoid collecting to a `list`.

jlebon commented on src/remote_prune:
```diff
@@ -0,0 +1,198 @@
+                        break
+                timestamp = build_id.split('.')[1]
+                buildDate = datetime.datetime(int(timestamp[0:4]), int(timestamp[4:6]), int(timestamp[-2:]))
+                if buildDate < refDate:
```
Can flip this and dedent.

Also, once we pass the point in the list where builds are no longer past the refData, we can just stop going backwards.

jlebon reviewed (commented):
Not a full review yet. Just a first pass.

jbtrystram commented:
We have been looking at this today with Adam and have the following thought : since all the clouds artifacts are stored in a S3 bucket, except the container images : should we split out the `builds.json` parsing to a shared library and have a separate script that uses that and do the quay API calls rather than writing non-s3 logic in there.

gursewak1997 commented:
> should we split out the `builds.json` parsing to a shared library and have a separate script that uses that and do the quay API calls rather than writing non-s3 logic in there ?

It would make sense to keep the overall pruning of builds/resources to be kept within this file. For the exact logic for deletion of resources, it's definitely okay to make a separate script; similar to what we have for [AWS](https://github.com/coreos/coreos-assembler/blob/main/src/cosalib/aws.py#L21) and [GCP](https://github.com/coreos/coreos-assembler/blob/main/src/cosalib/gcp.py#L21).
The S3 logic is just there to make sure the we get the information for `builds.json` and update the file accordingly once the pruning is done.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+    stream = args.stream
+    # Boto3 loads credentials from ~/.aws/config by default and we can change
+    # this default location by setting the AWS_CONFIG_FILE environment variable.
+    # The alternative is to manually pass ACCESS_KEY and SECRET_KEY which isn't favourable.
```
```suggestion
    # The Python bindings don't support passing a config file.
    # The alternative is to manually pass ACCESS_KEY and SECRET_KEY which isn't favourable.
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+    if stream not in policy:
+        print(f"No pruning policy specified for {stream} stream in policy.yaml")
+        return
+    else:
```
Delete and dedent everything under

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+        s3_client = boto3.client("s3")
+        # If the build key is set in the policy file, then the cloud-uploads key must
+        # also be present, and the duration of cloud-uploads must be equal or shorter
+        if "build" in policy[stream].keys():
```
Pretty sure you don't need `.keys()`.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+
+
+def cloud_uploads_check(actions):
+    if "cloud-uploads" in actions.keys():
```
Here neither.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+    if "cloud-uploads" in actions.keys():
+        cloud_uploads_duration = get_months(actions["cloud-uploads"])
+        build_duration = get_months(actions["build"])
+        assert cloud_uploads_duration < build_duration
```
```suggestion
        assert cloud_uploads_duration <= build_duration
```

jlebon replied:
Let's throw a proper exception instead with a clearer error message.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+            raise Exception(e)
+
+
+def cloud_uploads_check(actions):
```
I would reflow this whole function as:

```python
if 'cloud-uploads' not in actions:
	raise Exception(...)
cloud_uploads_duration = ...
build_duration = ...
if cloud_uploads_duration > build_duration:
	raise Exception(...)
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+    elif unit in ["months" or "month"]:
+        months = int(val)
+    else:
+        print("Duration of pruning for resources is only supported in years or months")
```
This should be the exception's text.

Also should include the problematic unit.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+
+def get_months(duration):
+    val, unit = duration.split(" ")
+    if unit in ["years" or "year"]:
```
```suggestion
    if unit in ["years", "year"]:
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+    val, unit = duration.split(" ")
+    if unit in ["years" or "year"]:
+        months = int(val) * 12
+    elif unit in ["months" or "month"]:
```
```suggestion
    elif unit in ["months", "month"]:
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+
+        # Base URL for Fedora CoreOS builds
+
+        buildJsonData = get_json_from_s3(bucket, prefix + "/builds/builds.json")
```
```suggestion
        builds_json_data = get_json_from_s3(bucket, prefix + "/builds/builds.json")
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+    parser.add_argument("--gcp-project", help="GCP Project name",
+                        default=os.environ.get("GCP_PROJECT_NAME"))
+
+    subparsers = parser.add_subparsers(dest="cmd", title="subcommands")
+    subparsers.required = True
+
+    s3 = subparsers.add_parser('s3', help='Prune s3 buckets')
+    s3.add_argument("url", metavar='<BUCKET>[/PREFIX]',
+                    help="Bucket and path prefix in which to upload")
+    s3.add_argument("--acl", help="ACL for objects",
+                    action='store', default='private')
+    s3.add_argument("--aws-config-file", default=os.environ.get("AWS_CONFIG_FILE"),
+                    help="Path to AWS config file")
```
Why is this a subparser?

gursewak1997 replied:
I took this from how we get s3 arguments in [buildupload](https://github.com/coreos/coreos-assembler/blob/main/src/cmd-buildupload#L57)

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+
+def parse_args():
+    parser = argparse.ArgumentParser(prog="coreos-assembler cloud-prune")
+    parser.add_argument("--policy", default='./policy.yaml', type=str,
```
Not sure this should have a default. To start, let's just always require it (use `required=True`).

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+def parse_args():
+    parser = argparse.ArgumentParser(prog="coreos-assembler cloud-prune")
+    parser.add_argument("--policy", default='./policy.yaml', type=str,
+                        help="Path to policy.yaml file")
```
```suggestion
                        help="Path to policy YAML file")
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+                        help="Path to policy.yaml file")
+    parser.add_argument("--dry-run", help="Don't actually delete anything",
+                        action='store_true')
+    parser.add_argument("--stream", type=str, help="Fedora stream", required=True)
```
```suggestion
    parser.add_argument("--stream", type=str, help="CoreOS stream", required=True)
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+        for action in policy[stream]:
+            duration = policy[stream][action]
+            duration_in_months = get_months(duration)
+            refDate = datetime.datetime.now() - relativedelta(months=int(duration_in_months))
```
```suggestion
            ref_date = datetime.datetime.now() - relativedelta(months=int(duration_in_months))
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+            duration_in_months = get_months(duration)
+            refDate = datetime.datetime.now() - relativedelta(months=int(duration_in_months))
+
+            print(f"Pruning {duration} old {action} for {stream} builds")
```
I think for debuggability, it'd be better to echo back what we parsed and calculated here. E.g.

```suggestion
            print(f"Pruning resources of type {action} older than {duration_in_months} months ({ref_date}) on stream {stream}")
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+            for index, build in enumerate(reversed(buildJsonData["builds"])):
+                build_id = build["id"]
+                if "policy-cleanup" in build.keys():
+                    # If we have already pruned the spedified resources for this
```
```suggestion
                    # If we have already pruned the specified resources for this
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+                    # If we have already pruned the spedified resources for this
+                    # build as per builds.json, we skip through it.
+                    if action in build["policy-cleanup"]:
+                        print(f"The {build_id} has already had {action} pruning completed")
```
```suggestion
                        print(f"Build {build_id} has already had {action} pruning completed")
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+                        print(f"The {build_id} has already had {action} pruning completed")
+                        continue
+                timestamp = build_id.split('.')[1]
+                buildDate = datetime.datetime(int(timestamp[0:4]), int(timestamp[4:6]), int(timestamp[-2:]))
```
Not super happy about this since it's of course FCOS-specific, and it doesn't feel great to auto-detect either. I think in the future we could have a new e.g. `--timestamp-source` that takes e.g. `fcos-version`, `rhcos-version`, or `build-meta`. The latter of which would actually fetch the `meta.json`.

Or... we enhance the `builds.json` schema to also have the build timestamp for each build. Really tempting...

Anyway, for now OK to assume FCOS versioning. I think let's factor this out into a function called e.g. `parse_fcos_version()`. It should parse the whole version string or bail if it can't. Use https://github.com/coreos/fedora-coreos-config/blob/testing-devel/versionary#L156 as a starting function definition and adapt that.

jlebon replied:
Also, `build_date`.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+                        continue
+                timestamp = build_id.split('.')[1]
+                buildDate = datetime.datetime(int(timestamp[0:4]), int(timestamp[4:6]), int(timestamp[-2:]))
+                if buildDate < refDate:
```
Invert and dedent everything under

gursewak1997 replied:
Can you elaborate this:)?

jlebon replied:
```python
if build_date >= ref_date:
    break
...
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+                buildDate = datetime.datetime(int(timestamp[0:4]), int(timestamp[4:6]), int(timestamp[-2:]))
+                if buildDate < refDate:
+                    for arch in build["arches"]:
+                        meta_url = f"/builds/{build_id}/{arch}/meta.json"
+                        meta_json = get_json_from_s3(bucket, prefix + meta_url)
```
How about `os.path.join()` instead?

gursewak1997 replied:
I am not sure how exactly that would help/get rid of the following check considering it's to access S3 directory.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+
+def get_json_from_s3(bucket, key):
+    # Create an S3 client
+    s3 = boto3.client("s3")
```
We already have a client we can pass through to this function.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+
+    except Exception as e:
+        print(f"Error fetching the JSON file from S3 {bucket}/{key}: {e}")
+        return None
```
Why aren't we reraising?

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+                                    buildJsonData["builds"][index] = build
+                            # Not implemented yet
+                            case "build":
+                                print(f"Deleting key {prefix}{build.id} from bucket {bucket}")
```
Shouldn't we raise a e.g. "Not implemented" exception here?

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+                    break
+        if not args.dry_run:
+            with open("builds/builds.json", "w") as json_file:
+                json_file.write(json.dumps(buildJsonData))
```
```suggestion
                json.dump(builds_json_data, json_file, indent=2)
```

jlebon replied:
We should also update the `timestamp` field.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+    s3 = boto3.client("s3")
+
+    try:
+        # Fetch the JSON file from S3
+        response = s3.get_object(Bucket=bucket, Key=key)
+
+        # Read the content of the file
+        content = response["Body"].read().decode("utf-8")
+
+        # Parse the JSON content
+        json_content = json.loads(content)
+
+        return json_content
+
+    except Exception as e:
```
Minor: seems overly spaced out in contrast to the rest of this script.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+# Handling just AWS and GCP at the moment
+def delete_cloud_resources(build, gcp_cloud_config, dry_run):
+    errors = []
+    totalBuildAMIs = len(build.images.get("amis") or "")
+    GCPImage = 0
+    # Unregister AMIs and snapshots
+    if not build.images.get("amis", []):
+        print(f"No AMIs for {build.id} for {build.arch}")
+    if not build.images.get("gcp", []):
+        print(f"No GCP image for {build.id} for {build.arch}")
+    else:
+        GCPImage = 1
```
I think instead of this, it'd be cleaner to have the dry run code be integrated into the logic below.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+            try:
+                deregister_ami(ami_id, region=region_name)
+            except Exception as e:
+                errors.append(e)
```
We need to safely handle the case where the AMI was already deregistered (i.e. it doesn't exist anymore) and not error out in that case. This is needed to handle rerunning safely if we're interrupted before we've re-uploaded the new `builds.json`.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+        snapshot_id = ami.get("snapshot")
+        if ami_id and region_name:
+            try:
+                deregister_ami(ami_id, region=region_name)
```
This is currently using the Python bindings. My inclination is to move away from that and just have it call `ore` instead, exactly like how `remove_gcp_image()` does it currently. For example, the golang code for deleting AMIs already exists and correctly handles the "is already deleted" case mentioned in my previous comment (see e.g. `deregisterImageIfExists()`).

That should also allow us to directly pass the config file to `ore` and avoid the env var hack.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+        gcp_image = gcp.get("image")
+        json_key = gcp_cloud_config.get("gcp", {}).get("json-key")
+        project = gcp_cloud_config.get("gcp", {}).get("project")
+        if gcp_image and json_key and project:
```
I feel like we should error out here if the information is incomplete.

gursewak1997 replied:
Do you mean in the case of either of the parameters (`gcp_image` and `json_key` and `project`) not being present.

jlebon replied:
Yes. We can't add the `cloud-uploads` marker to a build if we couldn't actually remove a cloud image just because we were missing information about how to remove it.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+
+    if len(errors) != 0:
+        print(f"Found errors when removing build {build.id}:")
+        for e in errors:
+            raise Exception(e)
```
That's not going to work as expected. On the first iteration, it will raise an exception which will cause the program to exit. I think here instead we would print all of the errors and then `raise` once.

gursewak1997 replied:
I see just printing the errors during the process as below. But how do you want to `raise` once? Would that be once we have gone through all the builds?
```
if len(errors) != 0:
    print(f"Found errors when removing cloud-uploads for {build.id}:")
    for e in errors:
        print(e)
```

jlebon replied:
You can add a `raise Exception("some errors were encountered")` right after the for-loop.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,303 @@
+    retry=retry_boto_exception,
+    retry_error_callback=retry_callback,
+)
+def s3_copy(s3_client, src, bucket, key, max_age, acl, extra_args={}, dry_run=False):
```
Let's move this to some library where we can share it with `cmd-buildupload`.

jlebon commented:
> Some thoughts while I was working on this :

Thanks for looking at this! I think from our discussions today, we landed on just keeping the container image pruning separate for now because (1) it seems more natural there to make the source of truth be the registry repo, (2) the GC policy is much more aggressive than the policy we have in mind here, so this would require adding another key to the `builds.json`'s `policy-cleanup` which would in principle need to be added to all builds we've created so far, and (3) there's extra logic needed there for barrier releases which makes it not as good a fit for the model here.

We could always look at re-merging them in the future if we feel it's worth it. I'll note this is similar also to the ostree repo pruner which treats the OSTree repo as canonical and is currently also handled by separate code (https://github.com/coreos/fedora-coreos-releng-automation/tree/main/fedora-ostree-pruner). I could similarly imagine having this pruning code be generalized to handle multiple repos under the Fedora umbrella (this relates to e.g. https://github.com/rpm-software-management/dnf5/issues/833#issuecomment-2000393754, but also we'll likely want tagged releases for the base bootc images too which will then also need GC).

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,272 @@
+        with open("builds/builds.json", "w") as json_file:
+            json.dump(builds_json_data, json_file, indent=2)
+        # This copies the local builds.json and updates the S3 bucket version.
+        if args.upload_builds_json:
```
Not quite. This needs to be a different mode entirely. So basically:
- `cosa cloud-prune --policy ...` prunes resources, updates `builds.json` locally.
- `cosa cloud-prune --upload-builds-json` does not do any pruning. It downloads the latest `builds.json`, merges the modifications from the local `builds.json` into the latest, and uploads to S3.

jlebon commented on src/cosalib/gcp.py:
```diff
@@ -26,6 +26,7 @@ def remove_gcp_image(gcp_id, json_key, project):
         ])
+        print(f"GCP: successfully removed image {gcp_id}")
     except SystemExit:
         raise Exception("Failed to remove image")
```
I guess for consistency with the AWS one let's add the `gcp_id` to this exception.

gursewak1997 commented on mantle/cmd/ore/gcloud/delete-images.go:
```diff
@@ -46,7 +47,14 @@ func runDeleteImage(cmd *cobra.Command, args []string) {
 		pending, err := api.DeleteImage(name)
 		if err != nil {
-			fmt.Fprintf(os.Stderr, "%v\n", err)
+			if gErr, ok := err.(*googleapi.Error); ok {
```
It was tricky to handle it from [api.DeleteImage()](https://github.com/coreos/coreos-assembler/blob/main/mantle/platform/api/gcloud/image.go#L223) considering how it handles the `pending` state. Will add the `--allow-missing` switch here though.

jlebon commented on mantle/cmd/ore/aws/delete-image.go:
```diff
@@ -0,0 +1,70 @@
+
+var (
+	cmdDeleteImage = &cobra.Command{
+		Use:   "delete-image --image <name>...",
```
Incorrect synopsis.

jlebon commented on mantle/cmd/ore/aws/delete-image.go:
```diff
@@ -0,0 +1,70 @@
+var (
+	cmdDeleteImage = &cobra.Command{
+		Use:   "delete-image --image <name>...",
+		Short: "Delete AMIs/Snapshots",
```
```suggestion
		Short: "Delete AMI and/or snapshot",
```

jlebon commented on mantle/cmd/ore/aws/delete-image.go:
```diff
@@ -0,0 +1,70 @@
+	AWS.AddCommand(cmdDeleteImage)
+	cmdDeleteImage.Flags().StringVar(&amiID, "ami", "", "AWS ami tag")
+	cmdDeleteImage.Flags().StringVar(&snapshotID, "snapshot", "", "AWS snapshot tag")
+	cmdDeleteImage.Flags().BoolVar(&allowMissing, "allow-missing", false, "Does not error out on the resource not existing")
```
```suggestion
	cmdDeleteImage.Flags().BoolVar(&allowMissing, "allow-missing", false, "Do not error out on the resource not existing")
```

jlebon commented on mantle/cmd/ore/aws/delete-image.go:
```diff
@@ -0,0 +1,70 @@
+	if amiID != "" {
+		err := API.RemoveByAmiTag(amiID, allowMissing)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "could not delete resource: %v\n", err)
```
The two error messages should be distinct; might as well also add the AMI ID/snapshot ID in there.

jlebon commented on mantle/cmd/ore/gcloud/delete-images.go:
```diff
@@ -29,10 +30,12 @@ var (
 
 func init() {
 	GCloud.AddCommand(cmdDeleteImage)
+	cmdDeleteImage.Flags().BoolVar(&allowMissing, "allow-missing", false, "Does not error out on the resource not existing")
```
```suggestion
	cmdDeleteImage.Flags().BoolVar(&allowMissing, "allow-missing", false, "Do not error out on the resource not existing")
```

jlebon commented on mantle/cmd/ore/gcloud/delete-images.go:
```diff
@@ -46,7 +49,15 @@ func runDeleteImage(cmd *cobra.Command, args []string) {
+			// Skip on NotFound error only if allowMissing flag is set to True
+			if gErr, ok := err.(*googleapi.Error); ok && allowMissing {
+				// Check if the error code is 404 (Not Found)
+				if gErr.Code == 404 {
```
`&& allowMissing` check should be here instead

jlebon replied:
Alternatively, do something similar what you did in the AWS one and only downcast in the `allow_missing` case.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,291 @@
+
+
+def parse_s3_url(builds_source_data_url):
+    if builds_source_data_url.startswith("s3://"):
+        bucket, prefix = builds_source_data_url.split('//', 1)[1].split('/', 1)
+        prefix = prefix[:-len("/builds")]
+        return bucket, prefix
+    else:
+        raise Exception("Invalid scheme: only s3:// supported")
```
This would be a good fit for `urllib.parse` instead: https://docs.python.org/3/library/urllib.parse.html

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,291 @@
+        builds_json_data["timestamp"] = datetime.datetime.now(pytz.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
+
+
+def parse_s3_url(builds_source_data_url):
```
```suggestion
def get_s3_bucket_and_prefix(url):
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,291 @@
+        if action not in policy[stream]:
+            continue
+        duration = get_period_in_months(policy[stream][action])
+        ref_date = datetime.datetime.now() - relativedelta(months=int(duration))
```
Minor: let's call `.now()` once outside the loop instead of multiple times. That will ensure we're consistently comparing against the same date across iterations.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,291 @@
+                images = get_supported_images(meta_json, unsupported, supported)
+                current_build = Build(id=build_id, images=images, arch=arch)
+
+                match action:
```
Missing case `images`.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,291 @@
+
+                match action:
+                    case "cloud-uploads":
+                        prune_cloud_uploads(current_build, cloud_config, build, builds_json_data, index, args.dry_run)
```
I think there's a bit of a layer violation here. We're passing way too much state to this function. Passing both a `Build` and also `build`, *and* also `index`, *and* `builds_json_data` seems off.

I think let's have the `Build` tuple include the meta JSON. Then `prune_cloud_uploads` should just take `current_build`, `cloud_config`, and `args.dry_run`. Then here after the arch for-loop, update the `policy-cleanup` of `build` (this fixes a bug in the current code: we don't want to update the `policy-cleanup` key on the first iteration, only once we've done all arches).

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,291 @@
+        # Save the updated builds.json to local builds/builds.json
+        save_builds_json(builds_json_data)
+
+    if args.upload_builds_json:
```
Again, this is a separate flow. I think let's just hoist this closer to the top of the function. So basically:

```python
def main():
    ...
    if args.upload_builds_json:
        return handle_upload_builds_json(...)
    # continue with regular GC
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,291 @@
+    if not dry_run:
+        build.setdefault("policy-cleanup", []).append("cloud-uploads")
+        builds_json_data["builds"][index] = build
+        builds_json_data["timestamp"] = datetime.datetime.now(pytz.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
```
We can do this once only at the end when we save the `builds.json` back to disk.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,291 @@
+
+
+def handle_upload_builds_json(s3_client, bucket, prefix, dry_run, acl):
+    remote_builds_json_data = get_json_from_s3(s3_client, bucket, os.path.join(prefix, "builds/builds.json"))
```
I think the prefix in this case will already have `.../builds`.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,291 @@
+    # Check if there are any changes that were made to remote(s3 version) builds.json
+    # while the pruning was in progress
+    if remote_builds_json_data != builds_json_source_data:
+        raise Exception("Detected remote modifications to builds.json")
```
We can't error out here. We need to implement the merge logic.

gursewak1997 replied:
Yep. This was just temporary. Need further discussion on the merge logic.

jlebon commented on src/cosalib/prune.py:
I think this PR mostly deprecates this file, except for the `get_unreferenced_s3_builds()` function. But all `delete_build()` I think we should just delete in the same commit where we delete `cosa remote-prune`.

gursewak1997 replied:
Yep, will delete this file when I split the commits. Changed this just for CI checks(`make check`) to pass.

jlebon replied:
OK with this, but as mentioned earlier, I think `get_unreferenced_s3_builds()` could still be useful. At least, I don't want to lose the overall idea behind it. I think basically `cosa cloud-prune` in the future should do this. Let's for now add a comment in `cmd-cloud-prune` like:

```python
# XXX: We should also prune unreferenced build directories here. See also
# `get_unreferenced_s3_builds()` in the git log.
```

jlebon reviewed (commented):
I still think it would be cleaner to have this commit split a bit more:
- the ore changes bits
- factoring out the S3 library functions
- adding the GC command
- removing remote-prune command

gursewak1997 commented:
Addressed the comments above except splitting it all into multiple relevant commits. 
Will do that once the code is good to go to make the splitting and committing easier till then.

jbtrystram commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,300 @@
+    return images
+
+
+def parse_fcos_version(version):
+    m = re.match(r'^([0-9]{2})\.([0-9]{8})\.([0-9]+)\.([0-9]+)$', version)
+    if m is None:
+        raise Exception(f"Incorrect versioning for FCOS build {version}")
+    try:
+        timestamp = datetime.datetime.strptime(m.group(2), '%Y%m%d')
+    except ValueError:
+        raise Exception(f"FCOS build {version} has incorrect date format. It should be in (%Y%m%d)")
+    return timestamp
```
This can be moved to cosalib to be shared with versionary and the containers pruning code

jlebon replied:
I was OK duplicating once but indeed now with the containers pruning code needing the same thing, it does make sense to put this in the library.

Not a blocker. We can do the move to a library in a follow-up or in the containers pruning PR.

jbtrystram commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,300 @@
+        s3_copy(s3_client, BUILDFILES['list'], bucket, f'{prefix}/builds.json', CACHE_MAX_AGE_METADATA, acl, extra_args={}, dry_run=dry_run)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(prog="coreos-assembler cloud-prune")
+    parser.add_argument("--policy", required=True, type=str, help="Path to policy YAML file")
+    parser.add_argument("--dry-run", help="Don't actually delete anything", action='store_true')
+    parser.add_argument("--upload-builds-json", help="Push builds.json", action='store_true')
+    parser.add_argument("--stream", type=str, help="CoreOS stream", required=True)
+    parser.add_argument("--gcp-json-key", help="GCP Service Account JSON Auth", default=os.environ.get("GCP_JSON_AUTH"))
+    parser.add_argument("--gcp-project", help="GCP Project name", default=os.environ.get("GCP_PROJECT_NAME"))
+    parser.add_argument("--acl", help="ACL for objects", action='store', default='private')
+    parser.add_argument("--aws-config-file", default=os.environ.get("AWS_CONFIG_FILE"), help="Path to AWS config file")
+    return parser.parse_args()
```
Can we move that towards the top ? When reading a script it's easier to have an idea of how it's meant to be used first

jlebon replied:
Agree with this, but I wouldn't consider it a blocker.

jbtrystram reviewed (changes requested):
Are we making sure we keep barrier releases ?

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,300 @@
+                        # print(f"Deleting key {prefix}{build.id} from bucket {bucket}")
+                        # Delete the build's directory in S3
+                        # S3().delete_object(args.bucket, f"{args.prefix}{str(current_build.id)}")
+                    case images:
```
"images"

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,300 @@
+        builds_json_source_data = json.load(f)
+    # Check if there are any changes that were made to remote(s3 version) builds.json
+    # while the pruning was in progress
+    if remote_builds_json_data != builds_json_source_data:
```
Minor: I would restructure this as:

```python
if ... == ...:
  s3_copy(...)
  return

# handle harder case here
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,300 @@
+        print("Detected remote updates to builds.json. Merging it to the local builds.json file")
+        with open(BUILDFILES['list'], "r") as f:
+            current_builds_json = json.load(f)
+        merged_builds_json = current_builds_json | remote_builds_json_data
```
I wasn't familiar with the `x | y` operator on dicts. But it looks like it's basically the same semantics as `x.update(y)` or so. Which is not quite what we want here. This will just take whatever `builds` is as a whole in `y` and not recurse into it.

We have `merge_dicts()` in cosalib, but it doesn't quite do what we need here either. We need something that treats the `id` member as the unique key.

We don't need a fancy merge algorithm here. I think to keep it simple, we can reduce this to: re-apply the `policy-cleanup` keys in `current_builds_json` to `remote_builds_json_data` (i.e. we don't need to recursively merge `policy-cleanup` lists, we can just always take the `policy-cleanup` key from `current_builds_json` since we know that wouldn't have changed). I'll let you doodle on the exact algorithm since it seems like a fun exercise. :)

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,300 @@
+def delete_gcp_image(build, cloud_config, dry_run):
+    errors = []
+    gcp = build.images.get("gcp")
+    if gcp:
```
Minor: can play the same trick here of handling the trivial case first and returning early, then dedent.

jlebon reviewed (commented):
This is getting close! I think the only main thing left is the merging the changes back into `builds.json` and then we're good to go.

gursewak1997 commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,300 @@
+            if not args.dry_run:
+                build.setdefault("policy-cleanup", []).append("cloud-uploads")
+                builds_json_data["builds"][index] = build
+                builds_json_data["timestamp"] = datetime.datetime.now(pytz.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
```
Do you mean outside this for loop because we do need the `index` to store the build data.

jlebon replied:
I mean moving just the timestamp bumping. We don't need to bump the timestamp multiple times. Just once, right before we write it out to disk.

gursewak1997 commented:
Worked on the minor changes and added the merging logic. Will split into multiple commits once gtg.

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,300 @@
+            current_builds_json = json.load(f)
+            update_policy_cleanup(current_builds_json, remote_builds_json)
+        if not dry_run:
+            # Make sure we have the merged json as local builds/builds.json
+            save_builds_json(remote_builds_json)
```
Ahh yes, good catch.

I guess in that case, since we're now uploading the same file either way, we can structure it this way:

```python
if remote_builds_json != builds_json_source_data:
    # do merge and update builds.json locally
# do s3_copy of local builds.json
```

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,300 @@
+
+
+# Function to update policy-cleanup keys into remote_builds
+def update_policy_cleanup(current_builds, remote_builds):
```
Nice, logic looks good!

jlebon commented on src/cmd-cloud-prune:
```diff
@@ -0,0 +1,297 @@
+            update_policy_cleanup(current_builds_json, remote_builds_json)
+        if not dry_run:
+            # Make sure we have the merged json as local builds/builds.json
+            save_builds_json(remote_builds_json)
```
We need to bump the timestamp here too.

Actually, probably simpler to just fold the timestamp bumping into `save_builds_json` directly.

jlebon reviewed (commented):
One comment, LGTM otherwise.
Let's split up the commit also.

jbtrystram reviewed (approved):
Superficial LGTM

jlebon reviewed (commented):
- Commit "Factor out the S3 functions to cosalib" needs to be *before* the introduction of `cmd-cloud-prune` which uses it.
- Let's add the name of the new command in the commit title. E.g.:

  > Add new `cosa cloud-prune` command for garbage collection

jlebon reviewed (approved):
LGTM. Nice work!
